sdk/core/azure-core/inc/azure/core/http/http.hpp

```diff
@@ -517,7 +517,7 @@ namespace Azure { namespace Core { namespace Http {
      *
      * @throw if \p name is an invalid header key.
      */
-    void AddHeader(std::string const& name, std::string const& value);
+    void SetHeader(std::string const& name, std::string const& value);
```

Comment:
The documentation says "Add HTTP header...". Why are we renaming this to Set?

`SetHeader` (or maybe even SetHeaderValue) only makes sense when we are overriding an **existing** header. It doesn't make sense when we want to Add a **new** header.

If I create a request with no headers, it is strange for me to start calling `SetHeader` to add new headers.

Reply:
Rather than renaming AddHeader, we should provide new SetHeader methods, if we want the "override" functionality.

Reply:
![image](https://user-images.githubusercontent.com/6527137/109354604-efb13c80-7832-11eb-8f0f-e06202b7248b.png)

Reply:
Our default behavior is to override headers when we call AddHeader().  We don't need the single AddHeader() which won't override if the key is already there (we don't want to throw or ignore the AddHeader() ).
We can either just document that our AddHeader will do override as if was SetHeader()  or we can document that SetHeader behaves similar to the `[]` operator from a map and it will add the key if it is not there or everride the value.

Reply:
@vhvb1989 What method do I call to add a **new** header to the request object?

cc @tg-msft

Reply:
>  which throws for existing keys

That's shooting ourselves in the foot =P 
A common case (probably the most important) is one Http policy that adds a header.  The policy doesn't care if the header is already there or not. It just want to override the value of it. So, it want to either add it or replace its value.

Right now, the policy just calls `AddHeader()` and if the header was already there, it gets overriden.

If we have both methods, the policy will need to check first if the header is there to know what method to call..  So, it making things complicated for us..  We only need one behavior, whichever the name is for the method.

I think SetHeader() make it clearer about the override.  But I would also be happy with `AddOrSetHeader()` or `AddOrOverrideHeader()`..  etc etc

Reply:
Both these options are good:
- `AddHeader()` + clear documentation that it will override existing ones OR 
- `AddOrSetHeader()`

The concern with just having `AddHeader`/`AddOrSetHeader`, is what happens if someone just wants to add a header, but doesn't want to override if the key already exists.